Mathlib/Algebra/Hom/Units.lean

```diff
@@ -512,3 +518,32 @@ protected theorem div_div_cancel_left (h : IsUnit a) : a / b / a = b⁻¹ := by
 end DivisionCommMonoid
 
 end IsUnit
+
+section IsLocalRingHom
+
+/-- A local ring homomorphism is a homomorphism `f` between monoids such that `a` in the domain
+  is a unit if `f a` is a unit for any `a`. See `LocalRing.local_hom_TFAE` for other equivalent
+  definitions in the local ring case - from where this concept originates, but it is useful in
+  other contexts, so we allow this generalisation in mathlib. -/
+class IsLocalRingHom [Monoid R] [Monoid S] [MonoidHomClass F R S] (f : F) : Prop where
+  /-- A local ring homomorphism `f : R ⟶ S` will send nonunits of `R` to nonunits of `S`. -/
+  map_nonunit : ∀ a, IsUnit (f a) → IsUnit a
+#align is_local_ring_hom IsLocalRingHom
+
+variable [Monoid R] [Monoid S] [Monoid T] [MonoidHomClass F R S]
+
+@[simp]
+theorem isUnit_map_iff (f : F) [IsLocalRingHom f] (a) : IsUnit (f a) ↔ IsUnit a :=
+  ⟨IsLocalRingHom.map_nonunit a, IsUnit.map f⟩
+#align is_unit_map_iff isUnit_map_iff
+
+@[simp]
+theorem IsUnit.of_map (f : F) [IsLocalRingHom f] (a) (h : IsUnit (f a)) : IsUnit a :=
+  IsLocalRingHom.map_nonunit a h
+#align is_unit_of_map_unit IsUnit.of_map
+
+theorem isLocalRingHom_of_leftInverse [MonoidHomClass G S R]
+  {f : F} (g : G) (hfg : Function.LeftInverse g f) : IsLocalRingHom f
+where map_nonunit a ha := by rwa [isUnit_map_of_leftInverse g hfg] at ha
```

Comment:
Mathlib4 style is
```suggestion
theorem isLocalRingHom_of_leftInverse [MonoidHomClass G S R]
    {f : F} (g : G) (hfg : Function.LeftInverse g f) : IsLocalRingHom f where
  map_nonunit a ha := by rwa [isUnit_map_of_leftInverse g hfg] at ha
```

Reply:
I'm not sure I fully get it, so there's double indent on the second line of proofs, and then single indent on proof, with the `where` after the type?